### DISCUSSION TITLE
Editing Designer Module Postprocessor?

### DISCUSSION THREAD

**Discussion by Trakyan** [winder/Universal-G-Code-Sender#2254] [Help]
``I was wondering if it's possible to add gcode snippets to the gcode generated by the designer module? Like to add a probe sequence or something before the cut starts or to move the M3 command to the start of the cut instead of the start of the file since for a CNC plasma/laser etc. having the tool switch on before it's in position. 

I watched the youtube video introducing the designer module by Breiler, but it didnt cover the post processor.

Below is a sample gcode generated with the designer plugin (the elipses are in place of the bulk of the gcode to save scrolling through it all, it was just a rectangle), I've added two comments at the points I'd like to modify. Does anyone know where the files/text/parameters that the plugin draws on to add those comments/gcode is stored so I can modify it?

`; This file was generated with "Universal Gcode Sender 2.0-SNAPSHOT / 2023-07-01"
;
; Tool: 3.0mm
; Depth per pass: 5.0mm
; Feed speed: 1000mm/min
; Plunge speed: 400mm/min
; Safe height: 5.0mm
; Tool step over: 0.3mm
; Spindle speed: 1000rpm

G21 ; millimeters
G90 ; absolute coordinate
G17 ; XY plane
G94 ; units per minute feed rate mode
M3 S1000 ; Turning on spindle <<HERE

; Rectangle - Outside path (1/1) <<HERE
G0 Z5
G0 X8.5Y38.5
G0 Z0
G1 F400 X8.5Y38.5Z-0
.......
G1 X8.5Y38.5Z-5
G0 Z5

; Turning off spindle
`M5`

Thanks

**Comment by breiler** [comment]
There is no way to set post processor parameters yet. Everything is pretty much hard coded here:
https://github.com/winder/Universal-G-Code-Sender/blob/master/ugs-platform/ugs-platform-plugin-designer/src/main/java/com/willwinder/ugs/nbp/designer/io/gcode/SimpleGcodeRouter.java#L156

I have been thinking of how to introduce settings for laser cutters but haven't gotten around to it yet. I don't have any experience with a plasma cutters but I expect it to behave similar to a laser cutter?

You probably would want to turn on the plasma torch just before any G1-command:
```
; Move into position
G0 Z5
G0 X8.5Y38.5
G0 Z0

; Cut shape
M3
G1 F400 X8.5Y38.5Z-0 S1000
......
G1 X8.5Y38.5Z-5
M5

; Move to next position
G0 Z5
```

**Reply by Trakyan** [reply]
Excellent, thank you! That looks somewhat similar to the post processor scripts i wrote for our CAM software, I'll have a look next week when I'm with the cnc computer at work.

For a plasma cutter the process is a bit more complex. On our machine at work it goes something like this:

-Torch moves to start of cut
-Torch probes to find material Z height (this is nessecary for every cut, since the material is never really flat.)
-Torch moves up to a pierce height
-Torch fires
-Torch pauses with G4 for some material and thickness dependant time to pierce all the way through.
-Torch moves down to cut height
-The torch height controller is activated and the machine starts cutting.

The CAM software we use has editable post possessing scripts that looks somewhat similar to what you linked. It has various functions available like for instance "on_set_feedrate()", "on_start_cut()" or whatever that get called whenever something specific happens. The user can then fill these functions with things like post.text("M3") or whatever they want to be done whenever that particular event occurs. I think something like this would allow for great versatility and flexibility for different machine types and setups (custom mcodes, routines etc.).

I can share the post processor script we use at work if that would be of any help as a reference or for ideas.

Edit:Having just had a look on my home computer, it seems that file is not accessible/editable from the download on the releases page. Does this mean I'd need to build it from source in order to edit that file?

**Reply by breiler** [reply]
I will gladly have a look at your post processing scripts to get some inspiration. I have been thinking of adding fields for gcode that can be executed when starting and finishing the job and scripts that will be executed before and after a cutting path. Would that work for you or does this require more advanced scripting?

The code that I linked to needs to be built from scratch.

**Reply by Trakyan** [reply]
[MartyBuilt Plasma Post.txt](https://github.com/winder/Universal-G-Code-Sender/files/12074087/MartyBuilt.Plasma.Post.txt)
That's the post processor i wrote for the plasma cutter at work. Sheetcam have pretty good documentation for how their post processors work. I was using one of the sample GRBL post processors they provide as a reference and starting point.
[GRBL plasma.txt](https://github.com/winder/Universal-G-Code-Sender/files/12074175/GRBL.plasma.txt)

There's also one with a plasma+scriber to show how it handles tool changes if that's helpful.
[GRBL THC with scriber.txt](https://github.com/winder/Universal-G-Code-Sender/files/12074179/GRBL.THC.with.scriber.txt)

[Post Documentation.txt](https://github.com/winder/Universal-G-Code-Sender/files/12074109/Post.Documentation.txt)
The above is the documentation sheetcam provides for the posts, The most interesting section is the first bit with the available events/functions. You can install the free/trial versionm of sheetcam (it only limits the length of gcode files you can generate if memory serves) to have a better look if you like, I copy/pasted it into a text document so it lost all the font and size data so it's a bit harder to read than in its original form. (it's under the toolbar options>machine then under the post processor tab there is a button "post documentation")

Hope that's useful in some way.

**Comment by Trakyan** [comment]
I forgot earlier, but here's a bit of sample Gcode generated with the "MartyBuilt Plasma Post" postprocessor, I figured that might be useful as well. It's just a square with rounded corners, and a "drilling" operation to mark a hole location in the center.
[Square Sample.txt](https://github.com/winder/Universal-G-Code-Sender/files/12074196/Square.Sample.txt)

I'd be happy to provide any more information you may find useful.
